<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>DecodeNcodeAnything: Designing DNA Storage in C++</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">DecodeNcodeAnything
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Designing DNA Storage in C++ </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><b>Decode Ncode Anything</b> <br  />
 <em>Caroline Leighton, Xinjin (Synthia) Li, Ryan Najac</em> <br  />
 <em>COMS 4995 Design using C++</em> <br  />
 <em>Fall 2023</em></p>
<h2><a class="anchor" id="autotoc_md1"></a>
1. Problems Addressed</h2>
<p >The exponential growth of the global datasphere necessitates a paradigm shift in storage solutions. Traditional methods like HDDs and SSDs are inadequate for the anticipated surge from 100 to 175 Zettabytes by 2025. This project aims to design an efficient and durable DNA storage system to meet the evolving demands of data preservation.</p>
<h2><a class="anchor" id="autotoc_md2"></a>
2. Alternatives Considered</h2>
<p >Current DNA storage technologies face challenges in short reads, long reads, and nanopore sequencing. Our design integrates Reed-Solomon error correction to enhance data reliability, addressing these challenges.</p>
<h2><a class="anchor" id="autotoc_md3"></a>
3. Final Design Overview</h2>
<p >Our approach involves opening, reading binary data, constructing sequences, and decoding. Key components include the Oligos class, <a class="el" href="classCodec.html" title="Codec class for handling files and Oligo data.">Codec</a> class, a vector for efficient data manipulation, and the <a class="el" href="classReedSolomon.html" title="Template class representing the Reed-Solomon error correction code.">ReedSolomon</a> template class for error correction.</p>
<h3><a class="anchor" id="autotoc_md4"></a>
Oligos Class: Efficient Storage and Retrieval</h3>
<p >The Oligos class efficiently stores DNA sequences using an unsigned 64-bit data block. It minimizes overhead by performing operations, such as slicing and conversion to a string representation, only upon request. This design choice optimizes for performance and resource usage.</p>
<p ><em>C++ Class Representation:</em> Utilizing a C++ class for Oligos aligns with the principles of efficiency and encapsulation. The use of an unsigned 64-bit data block allows for bitwise operations, enhancing speed and reducing memory footprint.</p>
<h3><a class="anchor" id="autotoc_md5"></a>
Codec Class: Managing File Stream and Operations</h3>
<p >The <a class="el" href="classCodec.html" title="Codec class for handling files and Oligo data.">Codec</a> class plays a crucial role in managing file-related operations, including opening, closing, and obtaining the filesize. This class encapsulates the complexities of interacting with the file stream, providing a streamlined interface for the rest of the system. The default constructor initializes the <a class="el" href="classCodec.html" title="Codec class for handling files and Oligo data.">Codec</a> object, while the constructor with a filename allows for easy instantiation with a specific file. The destructor ensures the proper closing of the file, enhancing resource management.</p>
<p ><em>C++ Class Representation:</em> Choosing a C++ class for the <a class="el" href="classCodec.html" title="Codec class for handling files and Oligo data.">Codec</a> concept offers a modular and organized approach to file handling. The class encapsulates file-related functionalities, promoting code readability and maintainability. The ability to easily open, close, and obtain file size through dedicated member functions enhances the overall robustness of the system. The class acts as a cohesive unit, abstracting away the intricacies of file stream management.</p>
<h3><a class="anchor" id="autotoc_md6"></a>
Reed-Solomon Error Correction: Ensuring Data Reliability</h3>
<p >In data transmission and storage, error correction is vital for data integrity. Reed-Solomon error correction involves adding redundant bits during encoding for error detection and correction during decoding.</p>
<h3><a class="anchor" id="autotoc_md7"></a>
Reed-Solomon Implementation with C++ Classes</h3>
<p >Our project seamlessly incorporates Reed-Solomon error correction using the <code><a class="el" href="classReedSolomon.html" title="Template class representing the Reed-Solomon error correction code.">ReedSolomon</a></code> template class. This class represents the Reed-Solomon error correction code, providing functions for encoding, decoding, and error correction.</p>
<ul>
<li><b>Galois Field:</b> The class incorporates a <code><a class="el" href="classGaloisField.html" title="Class representing a Galois Field of order 2^fieldPower.">GaloisField</a></code> attribute, fundamental to Reed-Solomon operations.</li>
</ul>
<h2><a class="anchor" id="autotoc_md8"></a>
4. Advantages of Our Algorithm</h2>
<ul>
<li><b>Efficient Storage and Retrieval:</b> Leveraging the Oligos class ensures faster bitwise operations.</li>
<li><b>Error Correction and Data Reliability:</b> Integration of Reed-Solomon enhances data reliability.</li>
<li><b>Flexibility and Compatibility:</b> Our design utilizes standard C++ libraries, ensuring compatibility.</li>
</ul>
<h2><a class="anchor" id="autotoc_md9"></a>
5. Tooling Choices for Efficient Development</h2>
<ul>
<li><b>CMake:</b> Used for building the project. CMake simplifies the build process, providing a platform-independent and efficient way to manage dependencies and compile the project seamlessly.</li>
<li><b>Doxygen:</b> Used for generating documentation from source code comments. Doxygen automates the documentation process, ensuring that code documentation remains up-to-date and accessible. This choice promotes clarity and understanding for developers interacting with the codebase.</li>
<li><b>zlib:</b> Important for handling .gz compressed files, especially since FASTQ sequencing files are generally compressed due to their size. zlib provides efficient and reliable compression and decompression algorithms, essential for working with large genomic datasets.</li>
</ul>
<h2><a class="anchor" id="autotoc_md10"></a>
6. Future Considerations</h2>
<p >Asynchronous Reading/Writing, a C++20 feature, could enhance performance. Given the directional nature of real-world data flows, it wasn't a primary focus.</p>
<h2><a class="anchor" id="autotoc_md11"></a>
7. Conclusion</h2>
<p >The proposed DNA storage design addresses current limitations, offering a robust, efficient, and reliable solution for the impending surge in global data. By combining innovative algorithms, well-designed classes, compatibility considerations, and strategic tooling choices, our project aims to contribute to the evolution of data storage technology.</p>
<hr  />
 </div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
