\hypertarget{md_README_autotoc_md22}{}\doxysubsection{Utilizing synthetic DNA ðŸ§¬ as a data storage ðŸ’¾ medium}\label{md_README_autotoc_md22}
A refactoring of the \href{https://github.com/lanl/adscodex}{\texttt{ Adaptive DNA Storage Codec (ADS Codex)}} in \char`\"{}modern\char`\"{} C++.\hypertarget{md_README_autotoc_md23}{}\doxysection{Table of Contents}\label{md_README_autotoc_md23}

\begin{DoxyItemize}
\item External Dependencies
\item Installation
\item Running the Program
\item Features
\item Documentation
\item Plans for Future Releases
\item Acknowledgments
\end{DoxyItemize}\hypertarget{md_README_autotoc_md24}{}\doxysection{External Dependencies}\label{md_README_autotoc_md24}

\begin{DoxyItemize}
\item {\bfseries{zlib for C++\+:}} a library for handling .gz files, necessary to support direct reading of compressed\+FASTQs. More information about zlib \href{https://www.zlib.net/}{\texttt{ here}}.
\item {\bfseries{CMake for Building\+:}} You can download CMake from \href{https://cmake.org/}{\texttt{ cmake.\+org}} or with your package manager.
\item {\bfseries{Doxygen for Documentation\+:}} used for automatically generating documentation from source code comments. More information about Doxygen can be found \href{https://www.doxygen.nl/}{\texttt{ here}}. \begin{quote}
Note that the Doxygen documentation for this project adheres to the style guide available \href{https://micro-os-plus.github.io/develop/doxygen-style-guide/}{\texttt{ here}}. \end{quote}

\end{DoxyItemize}\hypertarget{md_README_autotoc_md25}{}\doxysection{Installation}\label{md_README_autotoc_md25}

\begin{DoxyEnumerate}
\item {\bfseries{Clone the Decode\+Ncode\+Anything repository\+:}}


\begin{DoxyCode}{0}
\DoxyCodeLine{git clone https://github.com/rdnajac/DecodeNcodeAnything.git}
\DoxyCodeLine{cd DecodeNcodeAnything}

\end{DoxyCode}

\item {\bfseries{Create a build directory\+:}}


\begin{DoxyCode}{0}
\DoxyCodeLine{mkdir build}
\DoxyCodeLine{cd build}

\end{DoxyCode}

\item {\bfseries{Configure the project with CMake\+:}}

On Debian-\/based systems\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{cmake ..}

\end{DoxyCode}


On Windows (Make sure your Min\+GW-\/x64 version is newer than 11.\+0.\+0)\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{cmake -\/G "{}MinGW Makefiles"{} ..}

\end{DoxyCode}

\item {\bfseries{Build the project\+:}}


\begin{DoxyCode}{0}
\DoxyCodeLine{make}

\end{DoxyCode}

\end{DoxyEnumerate}

Adjust the build commands based on your specific build system or requirements. Alternatively, you can run the default installation script\+: {\ttfamily scripts/build.\+sh} from the repo\textquotesingle{}s top-\/level directory.\hypertarget{md_README_autotoc_md26}{}\doxysection{Documentation}\label{md_README_autotoc_md26}
To generate documentation using Doxygen, follow these steps\+:


\begin{DoxyEnumerate}
\item {\bfseries{Install Doxygen\+:}}

Ensure that Doxygen is installed on your system. If it\textquotesingle{}s not installed, you can typically install it using your package manager.

For example, on Debian-\/based systems\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{sudo apt-\/get install doxygen}

\end{DoxyCode}
 On Windows system\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{git clone https://github.com/doxygen/doxygen.git}

\end{DoxyCode}


Refer to the \href{https://www.doxygen.nl/manual/install.html}{\texttt{ Doxygen installation guide}} for more details.
\item {\bfseries{Navigate to the project root\+:}}


\begin{DoxyCode}{0}
\DoxyCodeLine{cd /path/to/DecodeNcodeAnything}

\end{DoxyCode}

\item {\bfseries{Run the documentation generation script\+:}}

On Debian-\/based systems


\begin{DoxyCode}{0}
\DoxyCodeLine{scripts/gen\_docs.sh}

\end{DoxyCode}


This script generates a new Doxyfile, configures it, and runs Doxygen to generate documentation in the {\ttfamily ./docs} folder.

On Windows system\+:

Just use Visual Studio to construct the project. Then it will generate a doxygen.\+exe.
\item {\bfseries{Access the documentation\+:}}

Open the generated documentation by navigating to the specified output directory\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{cd /path/to/DecodeNcodeAnything/docs}

\end{DoxyCode}


Open the {\ttfamily index.\+html} file in a web browser to explore the generated documentation.
\end{DoxyEnumerate}\hypertarget{md_README_autotoc_md27}{}\doxysection{Running the Program}\label{md_README_autotoc_md27}
After cloning and building the the program, the executables (including test programs) are located in the build directory. Running the program is simple\+: 
\begin{DoxyCode}{0}
\DoxyCodeLine{./build/app/encode <file-\/to-\/be-\/encoded>}

\end{DoxyCode}
 or alternatively, 
\begin{DoxyCode}{0}
\DoxyCodeLine{./build/app/encode <file-\/to-\/be-\/decoded>}

\end{DoxyCode}
 The decoder expects the FASTQ files while the encoder can handle any readable file.\hypertarget{md_README_autotoc_md28}{}\doxysection{Features}\label{md_README_autotoc_md28}
\hypertarget{md_README_autotoc_md29}{}\doxysubsection{Reedâ€“\+Solomon Error Correction}\label{md_README_autotoc_md29}
Library written in C++ for module export.

\begin{quote}
Reedâ€“\+Solomon Error Correction is a mathematical technique that allows the correction of errors in transmitted or stored data to enhance reliability and robustness. It is widely used in various applications, including data storage, QR codes, and digital communication. \end{quote}
Resources for understanding Reedâ€“\+Solomon error correction\+:
\begin{DoxyItemize}
\item \href{https://en.wikipedia.org/wiki/Reed\%E2\%80\%93Solomon_error_correction}{\texttt{ Reedâ€“\+Solomon Codes on Wikipedia}}
\item \href{https://www.cs.cmu.edu/~guyb/realworld/reedsolomon/reed_solomon_codes.html}{\texttt{ Introduction to Reedâ€“\+Solomon Codes}}
\end{DoxyItemize}\hypertarget{md_README_autotoc_md30}{}\doxysection{TODO\+: Implement $<$a href=\char`\"{}https\+://github.\+com/google/googletest\char`\"{} $>$\+Google\textquotesingle{}s testing framework$<$/a$>$.}\label{md_README_autotoc_md30}
Example\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{\textcolor{preprocessor}{\#include <gtest/gtest.h>}}
\DoxyCodeLine{}
\DoxyCodeLine{TEST(ADSCodexTest, EncodingTest) \{}
\DoxyCodeLine{    \textcolor{comment}{// Test encoding functionality of the ADS Codex}}
\DoxyCodeLine{    \textcolor{comment}{// ...}}
\DoxyCodeLine{    ASSERT\_TRUE(\textcolor{comment}{/* Some condition indicating success */});}
\DoxyCodeLine{\}}
\DoxyCodeLine{}
\DoxyCodeLine{TEST(ADSCodexTest, DecodingTest) \{}
\DoxyCodeLine{    \textcolor{comment}{// Test decoding functionality of the ADS Codex}}
\DoxyCodeLine{    \textcolor{comment}{// ...}}
\DoxyCodeLine{    ASSERT\_TRUE(\textcolor{comment}{/* Some condition indicating success */});}
\DoxyCodeLine{\}}
\DoxyCodeLine{}
\DoxyCodeLine{\textcolor{comment}{// Add more tests as needed...}}
\DoxyCodeLine{}
\DoxyCodeLine{\textcolor{keywordtype}{int} \mbox{\hyperlink{decode_8cpp_a0ddf1224851353fc92bfbff6f499fa97}{main}}(\textcolor{keywordtype}{int} argc, \textcolor{keywordtype}{char} **argv) \{}
\DoxyCodeLine{    ::testing::InitGoogleTest(\&argc, argv);}
\DoxyCodeLine{    \textcolor{keywordflow}{return} RUN\_ALL\_TESTS();}
\DoxyCodeLine{\}}

\end{DoxyCode}
\hypertarget{md_README_autotoc_md31}{}\doxysection{Plans for Future Releases}\label{md_README_autotoc_md31}

\begin{DoxyEnumerate}
\item {\bfseries{Performance Optimization with Lookup Tables\+:}}
\begin{DoxyItemize}
\item Introduce and leverage lookup tables for performance optimization. Lookup tables can enhance the efficiency of certain operations, contributing to faster encoding and decoding processes.
\end{DoxyItemize}
\item {\bfseries{Abstract Interface for \mbox{\hyperlink{classOligo}{Oligo}} Viability Criteria (H4\+G2)\+:}}
\begin{DoxyItemize}
\item Introduce an abstract interface for evaluating the viability of oligonucleotides based on specific criteria. One such criteria, denoted as H4\+G2, prevents the inclusion of oligos with homopolymers longer than 4 nucleotides (for A, T, and C) or 2 nucleotides for G. Developers can extend this interface to implement custom viability criteria.
\end{DoxyItemize}
\item {\bfseries{Expanded Documentation\+:}}
\begin{DoxyItemize}
\item Enhance and expand the documentation to provide comprehensive guidance on usage and potential extensions.
\end{DoxyItemize}
\end{DoxyEnumerate}

These plans are subject to change based on community feedback and project priorities. Stay tuned for updates and announcements related to future releases.

If you have specific features or improvements you would like to see in future releases, feel free to contribute to the discussion on our \href{https://github.com/rdnajac/DecodeNcodeAnything}{\texttt{ Git\+Hub repository}} or open a new issue.\hypertarget{md_README_autotoc_md32}{}\doxysection{Acknowledgments}\label{md_README_autotoc_md32}

\begin{DoxyItemize}
\item \href{https://github.com/lanl/adscodex}{\texttt{ Adaptive DNA Storage Codec (ADS Codex)}}\+: The foundation for this project.
\item \href{https://github.com/google/googletest}{\texttt{ Google Test}}\+: For providing a robust testing framework.
\item \href{https://cmake.org/}{\texttt{ CMake}}\+: Used for building the project.
\item \href{https://www.doxygen.nl/}{\texttt{ Doxygen}}\+: Used for generating documentation from source code comments.
\item \href{https://www.illumina.com/}{\texttt{ Illumina}}\+: For contributions to DNA sequencing technology.
\item \href{https://nanoporetech.com/}{\texttt{ Oxford Nanopore Technology}}\+: For advancements in nanopore sequencing.
\item \href{https://kilobaser.com/}{\texttt{ Kilobaser}}\+: For innovations in DNA synthesis technology.
\item \href{http://www.stroustrup.com/}{\texttt{ Bjarne Stroustrup}}\+: For his foundational contributions to C++ and for his guidance in this project. 
\end{DoxyItemize}